src/Typeahead/index.jsx

```diff
@@ -0,0 +1,287 @@
+const RESULTS = 'RESULTS'
+const TOO_MANY = 'TOO_MANY'
+const CHOSEN = 'CHOSEN'
+
+export const nextStateFromValue = (value, props) => {
+  let phase = value ? WAITING : EMPTY
+  if (value && value.length < props.minLength) {
+    phase = ACCUM
+  }
+
+  return {
+    inputValue: value,
+    phase,
+    searchResults: [],
+    selectedIndex: -1
+  }
+}
+
+export const nextStateFromOptions = (options, props) => {
+  let phase = RESULTS
+  if (!options || options.length === 0) {
+    phase = NO_RESULTS
+  }
+  else if ( options && options.length > props.maxVisible) {
+    phase = TOO_MANY
+  }
+
+  return {
+    phase,
+    searchResults: options,
+    selectedIndex: -1
+  }
+}
+
+// ----------------------------------------------------------------------------
+// Class
+
+export default class Typeahead extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = nextStateFromValue(props.value, props)
+    this.stateHistory = [this.state]
+
+    // Render/Phase Map
+    this.renderMap = {
+      ERROR: this._renderError.bind(this),
+      NOT_FOCUSED: this._renderEmpty.bind(this),
+      EMPTY: this._renderEmpty.bind(this),
+      ACCUM: this._renderEmpty.bind(this),
+      WAITING: this._renderWaiting.bind(this),
+      NO_RESULTS: this._renderNoResults.bind(this),
+      RESULTS: this._renderResults.bind(this),
+      TOO_MANY: this._renderTooManyResults.bind(this),
+      CHOSEN: this._renderEmpty.bind(this)
+    }
+
+    // Key/function map
+    this.keyMap = {}
+    this.keyMap[keys.VK_ESCAPE] = this._keyCancel.bind(this)
+    this.keyMap[keys.VK_UP] = this._nav.bind(this, -1)
+    this.keyMap[keys.VK_DOWN] = this._nav.bind(this, 1)
+
+    // TODO: If allowAnyValue psuedocode
+    // if (this.props.allowAnyValue) {
+    //   this.keyMap[keys.VK_ENTER] = this._keyEnter.bind(this)
+    //   this.keyMap[keys.VK_RETURN] = this._keyEnter.bind(this)
+    //   this.keyMap[keys.VK_TAB] = this._keyTab.bind(this)
+    //}
+    //else {
+    this.keyMap[keys.VK_ENTER] = this._chooseSelectedIndex.bind(this)
+    this.keyMap[keys.VK_RETURN] = this._chooseSelectedIndex.bind(this)
+    this.keyMap[keys.VK_TAB] = this._chooseSelectedIndex.bind(this)
+
+    // Bindings
+    this._onBlur = this._onBlur.bind(this)
+    this._onFocus = this._onFocus.bind(this)
+    this._onKeyDown = this._onKeyDown.bind(this)
+    this._setOptions = this._setOptions.bind(this)
+    this._valueUpdated = this._valueUpdated.bind(this)
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.setState(nextStateFromValue(nextProps.value, nextProps))
+  }
+
+  render() {
+    return (
+      <section className="typeahead"
+               onBlur={this._onBlur}
+               onFocus={this._onFocus}>
+        <input type="text"
+               disabled={this.props.disabled}
+               onChange={this._valueUpdated}
+               onKeyDown={this._onKeyDown}
+               placeholder={this.props.placeholder}
+               value={this.state.inputValue}
```

Comment:
Maybe here's a better place to set the top-level stuff and pass down so you can change it per distinct type-ahead instance (search Dispatch vs. Issue dispatch, etc)